lib/ex_doc/retriever.ex

```diff
@@ -382,14 +382,24 @@ defmodule ExDoc.Retriever do
     {{kind, name, arity}, anno, _, doc, metadata} = callback
     actual_def = actual_def(name, arity, kind)
     doc_line = anno_line(anno)
-    annotations = annotations_from_metadata(metadata)
 
-    {:attribute, anno, :callback, {^actual_def, specs}} =
-      Enum.find(module_data.abst_code, &match?({:attribute, _, :callback, {^actual_def, _}}, &1))
+    specs =
+      case Map.fetch(module_data.callbacks, actual_def) do
+        {:ok, specs} ->
+          specs
+
+        :error ->
+          # gen_statem:state_name/3 is stored as StateName/3 in the docs chunk
+          name = name |> Atom.to_string() |> Macro.underscore() |> String.to_atom()
```

Comment:
Is this a bug on the chunk?

Reply:
This is complicated.

We have:

```elixir
iex> :gen_statem.behaviour_info(:callbacks)
[
  state_name: 3,
  ...
]
```

And:

```elixir
Code.fetch_docs(:gen_statem) |> elem(6) |> Enum.map(&elem(&1, 0)) |> Enum.filter(&elem(&1, 0) == :callback)
[
  {:callback, :StateName, 3},
  # ...
]
```

And:

```elixir
iex> b :gen_statem
@callback state_name(:enter, oldStateName :: state_name(), data :: data()) ::
            state_enter_result(:state_name)
@callback state_name(event_type(), eventContent :: term(), data :: data()) ::
            event_handler_result(state_name())

# ...
```

And:

```elixir
b :gen_statem.state_name
@callback state_name(:enter, oldStateName :: state_name(), data :: data()) ::
            state_enter_result(:state_name)
@callback state_name(event_type(), eventContent :: term(), data :: data()) ::
            event_handler_result(state_name())
```

Note: no docs, even though they're properly attached to the chunk entry, because we have this discrepancy of `state_name` vs `StateName`.

Maybe it should be `state_name` in the chunk, that'd definitely solve the discrepancy, however looking at https://erlang.org/doc/man/gen_statem.html I think referring to it as `Module:StateName/3` was on purpose, both the  module name and function name vary.

If it makes sense to keep as `StateName` in the chunk, also worth mentioning that we won't be able to document such callback with EDoc given

```erlang
-callback StateName(Foo) -> ok.
```

won't compile. We'd probably need to add support for a `@name` tag to overwrite how it's stored in the chunk.

```erlang
-callback state_name(enter, OldState, Data).
%% @name StateName      (we document callbacks _after_ the definition.)
```

If we keep it as `state_name` in the chunk, maybe we do the reverse, store how the name should be displayed in chunk entry metadata.